### ISSUE TITLE
exp(p3): population-based training with mutation — basin escape via diversity

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#288]
## Thesis context

If #270 confirms basin trap (PPO can't reach cooperative basin from random init), PBT is the canonical solution: maintain a population of policies, mutate hyperparameters and weights, periodically replace bottom-performers with mutations of top-performers. Population diversity provides escape from local basins.

## Protocol

1. Maintain 8-16 PPO policy populations, each training in parallel on `minimal_specialization`
2. Every N iterations: evaluate, kill bottom 25%, replace with mutated copies of top 25%
3. Mutate: small perturbations to policy weights + hyperparameters (lr, entropy_coef)
4. Compare best-of-population gap_closed against single-policy PPO

## Verdict logic

| Outcome | Interpretation |
|---|---|
| Best-of-pop reaches gap_closed ≥ 0.7 | Population escape works. Cooperation is reachable; standard PPO just can't find it. |
| Population converges to defection (all 16) | Cooperative basin is *globally* unreachable from random init under SGD-style updates. |
| Slow convergence but eventually arrives | Pure escape problem, not gradient direction. Algorithmically uninteresting but operationally fine. |

## Effort

~3 days. Need to manage 8-16 parallel training instances + mutation logic. Significant compute (16× the single-PPO compute budget).

## Sequencing

Phase 2-3. After Phase 1 verdicts establish what we're solving for. PBT is most useful if basin trap; less useful if anti-attractor (PBT doesn't fix gradient direction, just helps find the right starting point).

## References

- Jaderberg et al. (2017) "Population Based Training of Neural Networks"
- `research_notebook/2026-05-17_thesis_misaligned_gradients.md`

---

### Status update

#270 has landed (closed) and #271's K=200 stability re-eval (PR #277, notebook `research_notebook/2026-05-17_issue271_best_of_n_verdict.md`) confirmed the **basin-trap reading**: PPO is locally trapped near the random-play basin, not actively repelled from cooperation. The basin-trap conclusion is what promotes this issue to Phase 2 load-bearing. The verdict matrix in the original body is now the live discriminator.

### Hypothesis

A population of independently-initialized PPO learners, periodically subjected to **selection + mutation**, will produce at least one individual whose `gap_closed` materially exceeds the single-PPO plateau on `minimal_specialization` (PPO plateau is gap_closed ≈ 0 from independent IPPO; specialist target is gap_closed ≈ 0.93 per #270/#272). The mechanism is population-level diversity that lets one lineage stumble close enough to the cooperative basin that subsequent gradient descent finds it — exactly what individual PPO from random init cannot do (#271).

If the population also collapses to defection across all 8-16 lineages, the cooperative basin is genuinely unreachable from random init under any local-search variant (Jaderberg-style PBT included). That is a *stronger* finding than the basin-trap reading: it suggests the gap between random-init basin and cooperative basin is too wide for diversity-plus-selection alone, and would re-elevate the gradient-direction interventions (#283 potential shaping, #287 LOLA, #284 COMA).

### PBT specification (Jaderberg et al. 2017)

| Knob | Value | Rationale |
|---|---|---|
| Population size | **16** (baseline), 8 (compute-constrained fallback) | Jaderberg uses 10-40; smaller is too noisy, larger blows the budget. 16 keeps total compute at 16× single-PPO. |
| Scenario | `minimal_specialization` (4 agents, 2 houses, ignition 0.05) | Same scenario as #270/#271 so verdict is directly comparable. |
| Per-individual training | 50 PPO iters between selection events, matching #270 cell length | Reuses existing `experiments/p3_specialization/train.py` per-cell harness. |
| Selection interval | Every **50 PPO iterations** (ready period in Jaderberg notation) | Long enough that policies differentiate; short enough to get 4-6 selection events per run. |
| Exploit step | Kill bottom 25% (4 of 16); each replaced by a **copy of a top-25% individual + mutation** | Truncation selection, standard PBT. |
| Total iterations | 300 PPO iters per lineage (6 selection events) | Budget cap; lineages that survive without replacement run the full 300. |
| **What mutates (weights)** | Add Gaussian noise σ=0.01 × layer-wise std to all policy + value head params | Small enough that one mutation doesn't break a working policy; large enough to break ties. |
| **What mutates (hyperparams)** | `lr` multiplied by uniform[0.8, 1.25]; `entropy_coef` multiplied by uniform[0.5, 2.0] | Jaderberg's perturbation factors. Mutate independently per knob with probability 0.5 each. |
| Selection criterion | Mean episode team reward over last 10 rollouts (≈ trailing-5 gap_closed proxy from `metrics.json`) | Same metric `analyze_270.py` uses for verdict classification. |
| Seeds (full PBT runs) | 3 independent PBT runs (seeds 42, 43, 44) | Need >1 PBT run for variance estimate; 3 is the minimum that lets us call a verdict. |

### Success criterion

Apply the same gap_closed thresholds as #270's verdict logic, but against the **best individual in the final population**:

| Best-of-population trailing-5 gap_closed (mean over 3 PBT runs) | Verdict |
|---|---|
| **≥ 0.7** | Population escape works. Cooperative basin reachable via diversity. Phase-2 success. |
| **0.2 – 0.7** | Partial escape. Population biases toward the basin but doesn't fully land. Suggests further iterations or larger population. |
| **< 0.2** | Cooperative basin globally unreachable from random init under local-search-with-diversity. Re-elevate gradient-direction interventions. |

**Comparison baseline**: Single-PPO IPPO on `minimal_specialization` lands at gap_closed ≈ 0 (see `analyze_270.py` reference table). Any best-of-pop above 0.2 is materially above the IPPO floor.

### Implementation guidance

Two reasonable build paths — Curator recommends Option A:

**Option A — thin PBT wrapper over the existing per-cell trainer (recommended)**

A new orchestrator script `experiments/p3_specialization/run_issue288_pbt.py` that:

1. Spawns 16 subprocesses, each calling `experiments/p3_specialization/train.py` with `--num-iterations 50 --output-dir <pop_root>/iter_0/lineage_<i>`, varying seed.
2. After each batch finishes, parses `metrics.json` from each cell, ranks lineages by trailing-5 mean team reward.
3. For bottom 25%: copies a randomly chosen top-25% lineage's `policies/agent_*.pt` checkpoint, perturbs weights, perturbs hyperparameters, and writes new `--bc-init-checkpoint-dir` + new `--lr --entropy-coef` flags.
4. Restarts all 16 lineages from their (possibly new) checkpoints for another 50 iters, repeating up to 6 selection events.

This reuses the validated training stack (joint trainer, normalizers, observation utils) and matches the file layout that `analyze_270.py` already understands. The orchestrator is the only new code (~200 LOC); the trainer is unchanged.

**Option B — embed PBT inside `bucket_brigade/training/population_trainer.py`**

`bucket_brigade/training/population_trainer.py:37-356` already coordinates multiple GPU learners over a shared simulator, but its semantics are *co-training in shared games*, not *independent lineages with selection*. Bolting selection/mutation onto it requires careful semantics around `matchmaking_strategy` and the multiprocessing queues, and the existing tests in `tests/test_population_training_integration.py` would need to extend. Higher risk; do not pursue first.

Independent of the build path, **all lineages should train on identical scenario seeds** within a generation so reward ranking reflects policy quality, not luck.

### Compute requirements (critical — see CLAUDE.md compute guidelines)

This experiment is **CPU-heavy and MUST run on a remote host**. Do NOT launch locally.

- Single-PPO cell on `minimal_specialization` (50 iters, 2048 rollout steps, 4 agents): ≈ 30-45 minutes per cell on `COMPUTE_HOST_PRIMARY` (Mac Studio class).
- 16 lineages × 6 generations = 96 cells per PBT run. With 8-way parallelism (matching `run_issue270_ppo_continuation.sh` pattern at `xargs -P 2`, scaled up for 16 lineages), ≈ 6-9 wall-clock hours per PBT run.
- 3 PBT runs (3 seeds) ≈ 18-27 wall-clock hours total. **Run in tmux** on the remote.

Launch pattern (per CLAUDE.md):

```bash
source .env
HOST="${COMPUTE_HOST_PRIMARY}"
ssh "$HOST"
cd ~/GitHub/bucket-brigade  # or ~/bucket-brigade
git pull
tmux new -s issue288_pbt
uv run python experiments/p3_specialization/run_issue288_pbt.py \
  --population-size 16 \
  --generations 6 \
  --iters-per-gen 50 \
  --seeds 42 43 44 \
  --output-dir experiments/p3_specialization/runs/issue288_pbt
# Ctrl+B, D to detach.
```

### Affected Files

- `experiments/p3_specialization/run_issue288_pbt.py` *(new)* — PBT orchestrator (Option A). Loops over generations, ranks lineages, applies exploit/explore, relaunches per-cell trainer.
- `experiments/p3_specialization/analyze_288.py` *(new)* — Verdict classifier, mirroring `analyze_270.py` structure. Reads each generation's per-lineage `metrics.json`, computes best-of-pop trajectory, applies the success-criterion table above, writes `results/issue288_pbt/{summary.md, summary.json}`.
- `experiments/p3_specialization/run_issue288_pbt.sh` *(new)* — Bash launcher mirroring `run_issue270_ppo_continuation.sh:1-40` pattern. Wraps the Python orchestrator with seed/output-dir defaults.
- `experiments/p3_specialization/train.py` *(touch)* — Confirm `--bc-init-checkpoint-dir`, `--lr`, `--entropy-coef` CLI args are exposed (they already are: `train.py:788`, see grep results). If a `--weight-perturbation-sigma` arg is needed for the explore step, it can be applied externally by mutating the checkpoint file before loading — no trainer change required.
- `research_notebook/2026-05-17_issue288_pbt_verdict.md` *(new at completion)* — Curator/Builder writes the verdict notebook after analysis, mirroring `research_notebook/2026-05-17_issue271_best_of_n_verdict.md`.

### Dependencies

- [x] #270 (basin-trap vs anti-attractor) — **closed**, basin-trap confirmed
- [x] #271 (random-init best-of-N vs PPO) — **closed**, PPO confirmed not worse than random
- [x] #272 (BC-fit alone) — **closed**, specialist representable

No open blockers. Issue is ready for human review after this curation pass.

### Related issues (do not block, but contextually adjacent)

- #284 COMA, #287 LOLA, #283 potential shaping — gradient-direction interventions. Demoted per #271 verdict, but become relevant again **if** #288 also fails (basin globally unreachable).
- #292 minimal 2x2 iterated dilemma — cleaner toy reproduction of basin trap. Useful sibling but independent of PBT.
- #291 single-agent long-horizon — orthogonal scope check.
- #275 Nash equilibrium of trained PPO — complementary analysis (game-theoretic vs algorithmic perspective).

### Test Plan

- [ ] **Smoke test (local, allowed)**: Run `run_issue288_pbt.py --population-size 4 --generations 2 --iters-per-gen 5 --seeds 42` end-to-end on a tiny scenario (or with `--num-agents 2`). Verify: directory layout matches `runs/issue288_pbt/seed_42/gen_{0,1}/lineage_{0..3}/`, mutation is applied to bottom 25% (1 of 4 lineages), `metrics.json` is written per cell, ranking logic picks the correct lineage as top. Wall-clock < 10 minutes locally.
- [ ] **Verdict classifier unit test**: Add a test that feeds synthetic per-lineage `metrics.json` files at the three threshold boundaries (best-of-pop gap_closed = 0.75, 0.45, 0.10) and asserts `analyze_288.py` returns `population_escape` / `partial` / `basin_globally_unreachable` respectively.
- [ ] **Full remote PBT run**: 3 seeds × 16 lineages × 6 generations × 50 iters on `COMPUTE_HOST_PRIMARY` per the launch pattern above. Acceptance: `analyze_288.py` produces `summary.md` with one of the three verdicts and emits `runs/issue288_pbt/seed_*/gen_*/lineage_*/metrics.json` for every cell.
- [ ] **Cross-check against #271 baseline**: Best-of-pop final gap_closed should be ≥ the K=200 single-best random-MLP result from #271 (gap_closed ≈ 0.122 independent, 0.146 shared). Falling below that floor would indicate a bug in the orchestrator, not a scientific finding.
- [ ] **Verdict notebook written**: `research_notebook/2026-05-17_issue288_pbt_verdict.md` documents the outcome and updates the queue priorities in `research_notebook/2026-05-17_issue271_best_of_n_verdict.md` (the post-#271 priority table).

### Scope guard

- This issue covers PBT *as a basin-escape probe*, not as a production training algorithm. Do not generalize beyond `minimal_specialization` or push the orchestrator into `bucket_brigade/training/` (Option B) within this PR.
- Do not blend PBT with gradient-direction interventions in the same run (no LOLA-PBT, no shaped-reward-PBT). Those are separate experiments if PBT fails.
- Estimated PR size: < 400 LOC (mostly orchestration glue + analysis). If the orchestrator drifts beyond this, file a decomposition follow-up rather than expanding scope here.

**Comment by rjwalters** [comment]
## Promoted to Phase 2 per #271 verdict

PR #277 confirmed the basin-trap reading: PPO is locally trapped in the random-play basin. Population-Based Training is **the canonical solution to local-basin trapping** via population diversity + mutation.

This issue is now one of the **most thesis-relevant** experiments. Previously Phase 3 'Hail Mary' — now Phase 2 load-bearing. If PBT escapes the basin, we have a complete demonstrated path: 'standard PPO trapped, PBT escapes.' If PBT also fails, the basin is genuinely unreachable from random init under any local-search variant — deeper finding.

See `research_notebook/2026-05-17_issue271_best_of_n_verdict.md`.

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete. Enhancement added as amended body (original preserved at top):

- Hypothesis tied to live #270/#271 verdicts (basin-trap confirmed)
- Concrete PBT spec table: 16 lineages, 50-iter ready period, 25% truncation, weight noise sigma=0.01, lr/entropy mutation factors per Jaderberg
- Success-criterion table mapped to gap_closed thresholds matching analyze_270.py verdict logic
- Affected Files lists new orchestrator (run_issue288_pbt.py), analyzer (analyze_288.py), launcher, verdict notebook
- Compute requirements explicit: CPU-heavy, MUST run remote via COMPUTE_HOST_PRIMARY + tmux (per CLAUDE.md)
- Implementation guidance recommends Option A (thin wrapper over existing experiments/p3_specialization/train.py) over Option B (extending bucket_brigade/training/population_trainer.py:37-356)
- Test Plan includes local smoke test, verdict-classifier unit test, full remote PBT run, cross-check against #271 best random-MLP floor (gap_closed approx 0.122/0.146)
- Scope guard caps PR at <400 LOC, no algorithm blending
- No open dependencies (#270/#271/#272 all closed). Ready for human approval (loom:issue).
